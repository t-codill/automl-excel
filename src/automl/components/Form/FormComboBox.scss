.ms-ComboBox-CaretDown-button {
    display: flex; 
    padding-left: 0;  
    .ms-Button-icon{
        height: 23px;
        margin-left: 11px;
    }    
}