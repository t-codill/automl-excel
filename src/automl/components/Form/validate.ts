import { FormDataType } from "./FormDataType";
import { IValidator } from "./Validators";

export function validate<TValue extends FormDataType>(value: TValue, validators: Array<IValidator<TValue>> | undefined): string | undefined {
    if (!validators) {
        return undefined;
    }
    for (const validator of validators) {
        const error = validator(value);
        if (error) {
            return error;
        }
    }
    return undefined;
}
