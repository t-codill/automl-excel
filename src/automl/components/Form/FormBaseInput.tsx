import * as React from "react";
import { Omit } from "../../common/Omit";
import { FormContext } from "./FormContext";
import { FormDataType } from "./FormDataType";
import { validate } from "./validate";
import { IValidator } from "./Validators";
export type FormBaseInputProps<
    TData extends { [key in TField]: TValue },
    TField extends keyof TData,
    TValue extends FormDataType,
    TProps> = Omit<TProps,
        "value"
        | "defaultValue"
        | "errorMessage"
        | "onChange"
        | "onChanged"
        | "defaultChecked"
        | "defaultValues"
        | "text"
        | "selectedKey"
        | "selectedKeys"
        | "defaultSelectedKey"
        | "defaultSelectedKeys"
        | "allowFreeform"
        | "validators"> & IFormBaseInputProps<TData, TField, TValue>;
export interface IFormBaseInputProps<
    TData extends { [key in TField]: TValue },
    TField extends keyof TData,
    TValue extends FormDataType
    > {
    field: TField;
    validators?: Array<IValidator<TValue>>;
    defaultFormValue?: TValue;
}
export interface IFormBaseInputStates<
    TData extends { [key in TField]: TValue },
    TField extends keyof TData,
    TValue extends FormDataType
    > {
    value: TValue;
    errorMessage: string | undefined;
}

export abstract class FormBaseInput<
    TData extends { [key in TField]: TValue },
    TField extends keyof TData,
    TValue extends FormDataType,
    TProps = {}
    >
    extends React.PureComponent<
    FormBaseInputProps<TData, TField, TValue, TProps>,
    IFormBaseInputStates<TData, TField, TValue>
    > {
    public static contextType = FormContext;

    public context!: React.ContextType<typeof FormContext>;

    constructor(props: FormBaseInputProps<TData, TField, TValue, TProps>) {
        super(props);
        this.state = {
            value: props.defaultFormValue as TValue,
            errorMessage: undefined
        };
    }

    public componentDidUpdate(prevProps: FormBaseInputProps<TData, TField, TValue, TProps>): void {
        if (prevProps.defaultFormValue !== this.props.defaultFormValue) {
            this.setState({ value: this.props.defaultFormValue as TValue });
        }
    }

    public componentDidMount(): void {
        this.context.mount(this);
    }

    public componentWillUnmount(): void {
        this.context.unmount(this);
    }

    public validate = () => {
        const error = validate<TValue>(this.state.value, this.props.validators);
        this.setState({ errorMessage: error });
        return error;
    }

    public getValue = (): TValue => {
        return this.state.value;
    }

    protected setValue(value: TValue): void {
        this.setState({
            value
        }, () => {
            if (this.context.onUpdated) {
                this.context.onUpdated(this.props.field, value);
            }
            this.validate();
        });
    }
}
