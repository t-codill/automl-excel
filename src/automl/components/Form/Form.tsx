import * as React from "react";
import { FormBaseInput } from "./FormBaseInput";
import { FormContext } from "./FormContext";
import { FormDataType } from "./FormDataType";

export interface IFormProps<TData extends { [key in keyof TData]: FormDataType }> {
    onSubmit(value: TData): void;
    onUpdated?(key: keyof TData, value: TData[keyof TData]): void;
}
export interface IFormStates<TData extends { [key in keyof TData]: FormDataType }> {
    onSubmit(value: TData): void;
}

export class Form<TData extends { [key in keyof TData]: FormDataType }> extends React.Component<IFormProps<TData>, IFormStates<TData>> {
    private readonly _inputs: Set<FormBaseInput<TData, keyof TData, TData[keyof TData]>> = new Set();
    public render(): React.ReactNode {
        return (
            <FormContext.Provider value={{
                mount: this._mount,
                unmount: this._unmount,
                onUpdated: this.props.onUpdated
            }}>
                <form
                    onSubmit={this._onSubmit}
                    noValidate={true}
                    autoComplete="off">
                    {this.props.children}
                </form>
            </FormContext.Provider>
        );

    }
    private readonly _mount = (input: FormBaseInput<TData, keyof TData, TData[keyof TData]>) => {
        this._inputs.add(input);
        return;
    }
    private readonly _unmount = (input: FormBaseInput<TData, keyof TData, TData[keyof TData]>) => {
        this._inputs.delete(input);
        return;
    }
    private readonly _onSubmit = (event: React.FormEvent<HTMLElement>): void => {
        event.preventDefault();
        const val: TData = Object.create(null);
        let valid = true;
        this._inputs.forEach((input) => {
            val[input.props.field] = input.getValue();
            valid = !input.validate() && valid;
        });
        if (valid) {
            this.props.onSubmit(val);
        }
    }
}
