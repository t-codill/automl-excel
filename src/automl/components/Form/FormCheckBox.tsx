import { Checkbox, ICheckboxProps } from "office-ui-fabric-react";
import * as React from "react";
import { FormBaseInput } from "./FormBaseInput";

export class FormCheckBox<
    TData extends { [key in TField]: boolean | undefined },
    TField extends keyof TData
    > extends FormBaseInput<TData, TField, boolean | undefined, ICheckboxProps>{
    public render(): React.ReactNode {
        return <Checkbox
            {...this.props}
            checked={this.state.value}
            onChange={this._onChange}
        />;
    }

    private readonly _onChange = (_?: React.FormEvent<HTMLElement | HTMLInputElement>, value?: boolean): void => {
        this.setValue(value);
    }
}
