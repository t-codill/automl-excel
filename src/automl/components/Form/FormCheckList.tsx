import * as React from "react";
import { CheckList, ICheckListProps } from "../CheckList/CheckList";
import { FormBaseInput } from "./FormBaseInput";

export class FormCheckList<
  TData extends { [key in TField]: string[] },
  TField extends keyof TData
  > extends FormBaseInput<TData, TField, string[], ICheckListProps>{
  public render(): React.ReactNode {
    return (
      <CheckList
        items={[]}
        {...this.props}
        onChange={this._onChange}
        values={this.state.value}
        errorMessage={this.state.errorMessage}
      />
    );
  }

  private readonly _onChange = (value: string[]): void => {
    this.setValue(value);
  }
}
