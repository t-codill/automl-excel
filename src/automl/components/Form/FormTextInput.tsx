import { ITextFieldProps, TextField } from "office-ui-fabric-react";
import * as React from "react";
import { FormBaseInput } from "./FormBaseInput";

export class FormTextInput<
    TData extends { [key in TField]: string | undefined },
    TField extends keyof TData
    > extends FormBaseInput<TData, TField, string | undefined, ITextFieldProps>{
    public render(): React.ReactNode {
        return <TextField
            {...this.props}
            value={this.state.value}
            onChange={this._onChange}
            errorMessage={this.state.errorMessage}
        />;
    }

    private readonly _onChange = (_: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, value?: string): void => {
        this.setValue(value);
    }
}
