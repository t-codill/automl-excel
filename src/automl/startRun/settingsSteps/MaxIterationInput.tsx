import * as React from "react";
import { FormTextInput } from "../../components/Form/FormTextInput";
import { Validators } from "../../components/Form/Validators";
import { LabelWithCallout } from "../../components/Label/LabelWithCallout";
import { ISettingsStepParams } from "./ISettingsStepParams";

const maxIterationInput: React.FunctionComponent = () => {
    const inputId = "maxIterationInput";
    return <div className="ms-Grid" dir="ltr">
        <div className="ms-Grid-col ms-sm12 ms-md6 ms-lg6 ms-xl6">
            <LabelWithCallout required={false} htmlFor={inputId} labelText="Max number of iterations">
                Max number of pipelines (algorithm and hyperparameters) to try.
            </LabelWithCallout>
        </div>
        <div className="ms-Grid-col ms-sm12 ms-md6 ms-lg6 ms-xl6">
            <FormTextInput<ISettingsStepParams, "maxIteration">
                field="maxIteration"
                placeholder="Max number of iteration"
                ariaLabel="Please enter training job time here"
                validators={[
                    Validators.isInteger("Max number of iterations should be an integer"),
                    Validators.minValue(1, () => "Max number of iterations should be greater than or equal to 1"),
                    Validators.maxValue(1000, () => "Max number of iterations cannot exceed 1000")
                ]}
                defaultFormValue="100"
            />
        </div>
    </div>;
};

export { maxIterationInput as MaxIterationInput };
