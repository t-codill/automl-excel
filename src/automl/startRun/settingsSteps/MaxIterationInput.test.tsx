import { shallow } from "enzyme";
import * as React from "react";
import { FormTextInput } from "../../components/Form/FormTextInput";
import { validate } from "../../components/Form/validate";
import { MaxIterationInput } from "./MaxIterationInput";

describe("MaxIterationInput", () => {
    it("should render", () => {
        const tree = shallow(
            <MaxIterationInput />
        );
        expect(tree)
            .toMatchSnapshot();
    });

    it("should validate", () => {
        const tree = shallow(
            <MaxIterationInput />
        );
        expect(validate(
            "1",
            tree
                .find(FormTextInput)
                .prop("validators")))
            .toBeUndefined();
    });

    it("should error for 0", () => {
        const tree = shallow(
            <MaxIterationInput />
        );
        expect(validate(
            "0",
            tree
                .find(FormTextInput)
                .prop("validators")))
            .toBe("Max number of iterations should be greater than or equal to 1");
    });

    it("should error for anything larger than 1000", () => {
        const tree = shallow(
            <MaxIterationInput />
        );
        expect(validate(
            "1001",
            tree
                .find(FormTextInput)
                .prop("validators")))
            .toBe("Max number of iterations cannot exceed 1000");
    });
});
