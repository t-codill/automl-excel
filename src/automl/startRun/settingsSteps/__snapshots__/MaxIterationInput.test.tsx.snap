// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaxIterationInput should render 1`] = `
<div
  className="ms-Grid"
  dir="ltr"
>
  <div
    className="ms-Grid-col ms-sm12 ms-md6 ms-lg6 ms-xl6"
  >
    <LabelWithCallout
      htmlFor="maxIterationInput"
      labelText="Max number of iterations"
      required={false}
    >
      Max number of pipelines (algorithm and hyperparameters) to try.
    </LabelWithCallout>
  </div>
  <div
    className="ms-Grid-col ms-sm12 ms-md6 ms-lg6 ms-xl6"
  >
    <FormTextInput
      ariaLabel="Please enter training job time here"
      defaultFormValue="100"
      field="maxIteration"
      placeholder="Max number of iteration"
      validators={
        Array [
          [Function],
          [Function],
          [Function],
        ]
      }
    />
  </div>
</div>
`;
